### PR TITLE
Feat/copy

### PR DESCRIPTION
**Pull request by BotellaA**

### CODE REVIEW

File: src/geode/basic/attribute_manager.cpp [hunk 1]
```diff
@@ -134,6 +134,32 @@ namespace geode
...
+                if( it != attributes_.end() )
+                {
+                    try
+                    {
+                        it->second->copy( *attribute.second, nb_elements_ );
+                    }
+                    catch( const std::bad_cast &e )
+                    {
+                        Logger::error( attribute.first,
+                            " attribute cannot be copied: ", e.what() );
```
Comment (panquez) [inline src/geode/basic/attribute_manager.cpp]: I propose:
"Attribute \\"", attribute.first, "\\" cannot be copied: "

File: src/geode/mesh/core/edged_curve.cpp [hunk 2]
```diff
@@ -87,6 +88,16 @@ namespace geode
         archive.ext( *this, bitsery::ext::BaseClass< Graph >{} );
     }
 
+    template < index_t dimension >
+    std::unique_ptr< EdgedCurve< dimension > >
+        EdgedCurve< dimension >::clone() const
+    {
+        auto new_clone = create( type_name() );
```
Comment (panquez) [inline src/geode/mesh/core/edged_curve.cpp]: I propose to rename "new_clone" by "clone"

File: src/geode/mesh/core/tetrahedral_solid.cpp [hunk 3]
```diff
@@ -53,5 +54,16 @@ namespace geode
         }
     }
```
Comment (panquez) [inline src/geode/mesh/core/tetrahedral_solid.cpp]: How to extend what you have changed for PolyhedralSolid here? Imagine a dependant project where are defined Solids inherited from TetrahedralSolid, how to do this recursive tries/catches?
Actually same question for PolyhedralSolid

File: tests/mesh/test-edged-curve.cpp [hunk 4]
```diff
@@ -157,6 +158,26 @@ void test_edge_requests( const geode::EdgedCurve3D& edged_curve,
         "[Test] Edge length is not correct" );
 }
 
+void test_copy( const geode::EdgedCurve3D& edged_curve )
```
Comment (panquez) [inline tests/mesh/test-edged-curve.cpp]:
```suggestion
void test_clone( const geode::EdgedCurve3D& edged_curve )
```

File: tests/mesh/test-edged-curve.cpp [hunk 5]
```diff
@@ -157,6 +158,26 @@ void test_edge_requests( const geode::EdgedCurve3D& edged_curve,
...
+
+    auto attribute2 =
+        edged_curve2->edge_attribute_manager().find_attribute< int >( "test" );
+    OPENGEODE_EXCEPTION( attribute2->value( 0 ) == 42,
+        "[Test] EdgedCurve2 attribute should be 42" );
+}
+
 int main()
 {
     using namespace geode;
```
Comment (panquez) [inline tests/mesh/test-edged-curve.cpp]: You forget to call test_clone()

File: tests/mesh/test-graph.cpp [hunk 6]
```diff
@@ -22,6 +22,7 @@
  */
 
 #include <geode/basic/attribute.h>
+#include <geode/basic/attribute_manager.h>
```
Comment (panquez) [inline tests/mesh/test-graph.cpp]: Remove these 2 includes about attributes not used here

File: tests/mesh/test-graph.cpp [hunk 7]
```diff
@@ -120,13 +121,19 @@ void test_delete_edge( const geode::Graph& graph, geode::GraphBuilder& builder )
...
+    builder.set_edge_vertex( { 1, 1 }, 0 );
     OPENGEODE_EXCEPTION(
         graph.nb_edges() == 11, "[Test] Graph should have 11 edges" );
 
     to_delete.resize( graph.nb_edges(), true );
     builder.delete_edges( to_delete );
     OPENGEODE_EXCEPTION(
         graph.nb_edges() == 1, "[Test] Graph should have 1 edge" );
+    OPENGEODE_EXCEPTION( graph.edge_vertex( { 0, 0 } ) == 1,
+        "[Test] Graph edge vertex index is not correct" );
```
Comment (panquez) [inline tests/mesh/test-graph.cpp]:
```suggestion
        "[Test] Graph edge vertex index is not correct (0, 0)" );
```

File: tests/mesh/test-graph.cpp [hunk 8]
```diff
@@ -120,13 +121,19 @@ void test_delete_edge( const geode::Graph& graph, geode::GraphBuilder& builder )
...
+    OPENGEODE_EXCEPTION( graph.edge_vertex( { 0, 1 } ) == 0,
+        "[Test] Graph edge vertex index is not correct" );
```
Comment (panquez) [inline tests/mesh/test-graph.cpp]:
```suggestion
        "[Test] Graph edge vertex index is not correct (0,1)" );
```
The objective is to have different error messages to quickly differentiate that is wrong

File: tests/mesh/test-graph.cpp [hunk 9]
```diff
@@ -139,6 +146,15 @@ void test_io( const geode::Graph& graph, const std::string& filename )
     load_graph( *new_graph, filename );
 }
 
+void test_copy( const geode::Graph& graph )
```
Comment (panquez) [inline tests/mesh/test-graph.cpp]:
```suggestion
void test_clone( const geode::Graph& graph )
```

File: tests/mesh/test-graph.cpp [hunk 10]
```diff
@@ -155,6 +171,7 @@ int main()
 
         test_delete_vertex( *graph, *builder );
         test_delete_edge( *graph, *builder );
+        test_copy( *graph );
```
Comment (panquez) [inline tests/mesh/test-graph.cpp]:
```suggestion
        test_clone( *graph );
```

File: tests/mesh/test-point-set.cpp [hunk 11]
```diff
@@ -84,6 +84,18 @@ void test_io( const geode::PointSet3D& point_set, const std::string& filename )
     load_point_set( *new_point_set, filename );
 }
 
+void test_copy( const geode::PointSet3D& point_set )
```
Comment (panquez) [inline tests/mesh/test-point-set.cpp]:
```suggestion
void test_clone( const geode::PointSet3D& point_set )
```

File: tests/mesh/test-point-set.cpp [hunk 12]
```diff
@@ -98,6 +110,7 @@ int main()
         auto base_file = "test." + point_set->native_extension();
         test_io( *point_set, base_file );
         test_delete_vertex( *point_set, *builder );
+        test_copy( *point_set );
```
Comment (panquez) [inline tests/mesh/test-point-set.cpp]:
```suggestion
        test_clone( *point_set );
```

File: tests/mesh/test-polygonal-surface.cpp [hunk 13]
```diff
@@ -271,6 +271,25 @@ void test_io( const geode::PolygonalSurface3D& polygonal_surface,
     load_polygonal_surface( *new_polygonal_surface, filename );
 }
 
+void test_copy( const geode::PolygonalSurface3D& polygonal_surface )
```
Comment (panquez) [inline tests/mesh/test-polygonal-surface.cpp]:
```suggestion
void test_clone( const geode::PolygonalSurface3D& polygonal_surface )
```

File: tests/mesh/test-polygonal-surface.cpp [hunk 14]
```diff
@@ -298,6 +317,7 @@ int main()
 
         test_delete_vertex( *polygonal_surface, *builder );
         test_delete_polygon( *polygonal_surface, *builder );
+        test_copy( *polygonal_surface );
```
Comment (panquez) [inline tests/mesh/test-polygonal-surface.cpp]:
```suggestion
        test_clone( *polygonal_surface );
```

File: tests/mesh/test-polyhedral-solid.cpp [hunk 15]
```diff
@@ -162,6 +163,38 @@ void test_barycenters()
...
+        polyhedral_solid.vertex_attribute_manager()
+            .find_or_create_attribute< geode::VariableAttribute, double >(
+                "test" );
+    for( auto v : geode::Range{ polyhedral_solid.nb_vertices() } )
+    {
+        attribute->value( v ) = v;
+    }
+}
+
+void test_copy( const geode::PolyhedralSolid3D& polyhedral_solid )
```
Comment (panquez) [inline tests/mesh/test-polyhedral-solid.cpp]:
```suggestion
void test_clone( const geode::PolyhedralSolid3D& polyhedral_solid )
```

File: tests/mesh/test-polyhedral-solid.cpp [hunk 16]
```diff
@@ -173,13 +206,15 @@ int main()
...
         test_create_vertices( *polyhedral_solid, *builder );
+        test_create_vertex_attribute( *polyhedral_solid );
         test_create_polyhedra( *polyhedral_solid, *builder );
         test_polyhedron_adjacencies( *polyhedral_solid, *builder );
         auto base_file = "test." + polyhedral_solid->native_extension();
         test_io( *polyhedral_solid, base_file );
 
         test_delete_vertex( *polyhedral_solid, *builder );
         test_delete_polyhedra( *polyhedral_solid, *builder );
+        test_copy( *polyhedral_solid );
```
Comment (panquez) [inline tests/mesh/test-polyhedral-solid.cpp]:
```suggestion
        test_clone( *polyhedral_solid );
```

File: tests/mesh/test-tetrahedral-solid.cpp [hunk 17]
```diff
@@ -118,6 +118,15 @@ void test_io(
     load_tetrahedral_solid( *new_solid, filename );
```
Comment (panquez) [inline tests/mesh/test-tetrahedral-solid.cpp]: remove include attribute

File: tests/mesh/test-tetrahedral-solid.cpp [hunk 18]
```diff
@@ -118,6 +118,15 @@ void test_io(
...
 }
 
+void test_copy( const geode::TetrahedralSolid3D& solid )
```
Comment (panquez) [inline tests/mesh/test-tetrahedral-solid.cpp]:
```suggestion
void test_clone( const geode::TetrahedralSolid3D& solid )
```

File: tests/mesh/test-tetrahedral-solid.cpp [hunk 19]
```diff
@@ -136,6 +145,7 @@ int main()
 
         test_delete_vertex( *solid, *builder );
         test_delete_polyhedron( *solid, *builder );
+        test_copy( *solid );
```
Comment (panquez) [inline tests/mesh/test-tetrahedral-solid.cpp]:
```suggestion
        test_clone( *solid );
```

File: tests/mesh/test-triangulated-surface.cpp [hunk 20]
```diff
@@ -114,6 +114,15 @@ void test_io(
     load_triangulated_surface( *new_surface, filename );
 }
```
Comment (panquez) [inline tests/mesh/test-triangulated-surface.cpp]: check and remove include attribute if unnecessary

File: tests/mesh/test-triangulated-surface.cpp [hunk 21]
```diff
@@ -114,6 +114,15 @@ void test_io(
...
 
+void test_copy( const geode::TriangulatedSurface3D& surface )
```
Comment (panquez) [inline tests/mesh/test-triangulated-surface.cpp]:
```suggestion
void test_clone( const geode::TriangulatedSurface3D& surface )
```

File: tests/mesh/test-triangulated-surface.cpp [hunk 22]
```diff
@@ -132,6 +141,7 @@ int main()
 
         test_delete_vertex( *surface, *builder );
         test_delete_polygon( *surface, *builder );
+        test_copy( *surface );
```
Comment (panquez) [inline tests/mesh/test-triangulated-surface.cpp]:
```suggestion
        test_clone( *surface );
```

File: tests/mesh/test-vertex-set.cpp [hunk 23]
```diff
@@ -78,6 +78,10 @@ int main()
         test_create_vertices( vertex_set, builder );
         test_delete_vertex( vertex_set, builder );
 
+        auto vertex_set2 = vertex_set.clone();
```
Comment (panquez) [inline tests/mesh/test-vertex-set.cpp]: put this test in a function

File: tests/model/test-brep.cpp [hunk 24]
```diff
@@ -515,6 +515,112 @@ void test_reloaded_brep( const geode::BRep& model )
...
+    auto mapping = builder.copy_components( brep );
+    builder.copy_component_relationships( mapping, brep );
+    OPENGEODE_EXCEPTION(
+        brep2.nb_corners() == 12, "[Test] BRep should have 12 corners" );
+    OPENGEODE_EXCEPTION(
+        brep2.nb_lines() == 18, "[Test] BRep should have 18 lines" );
+    OPENGEODE_EXCEPTION(
+        brep2.nb_surfaces() == 10, "[Test] BRep should have 10 surfaces" );
+    OPENGEODE_EXCEPTION(
+        brep2.nb_blocks() == 2, "[Test] BRep should have 2 block" );
```
Comment (panquez) [inline tests/model/test-brep.cpp]:
```suggestion
        brep2.nb_blocks() == 2, "[Test] BRep should have 2 blocks" );
```

File: tests/model/test-brep.cpp [hunk 25]
```diff
@@ -515,6 +515,112 @@ void test_reloaded_brep( const geode::BRep& model )
...
+    OPENGEODE_EXCEPTION( brep2.nb_model_boundaries() == 6,
+        "[Test] BRep should have 6 model boundaries" );
+
+    for( const auto& corner : brep.corners() )
+    {
+        const auto& new_corner =
+            brep2.corner( mapping.corners.at( corner.id() ) );
+        for( const auto& line : brep.incidences( corner ) )
+        {
+            bool found = { false };
```
Comment (panquez) [inline tests/model/test-brep.cpp]: {} with = ?

File: tests/mesh/test-point-set.cpp [hunk 26]
```diff
@@ -84,6 +84,18 @@ void test_io( const geode::PointSet3D& point_set, const std::string& filename )
...
+void test_copy( const geode::PointSet3D& point_set )
+{
+    auto point_set2 = point_set.clone();
+    OPENGEODE_EXCEPTION( point_set2->nb_vertices() == 3,
+        "[Test] PointSet2 should have 3 vertices" );
+
+    auto attribute =
+        point_set2->vertex_attribute_manager().find_attribute< bool >( "test" );
+    OPENGEODE_EXCEPTION( attribute->value( 0 ) == true,
+        "[Test] PointSet2 attribute value should be true" );
```
Comment (panquez) [inline tests/mesh/test-point-set.cpp]: I suggest to add a test on coordinates values, using this answer:
geode::Point3D answer{ { 2.1, 9.4, 6.7 } };

File: src/geode/mesh/core/tetrahedral_solid.cpp (hunk 3 above)
Comment (BotellaA) [reply src/geode/mesh/core/tetrahedral_solid.cpp]: You can't... but should you create a new type outside of OG? A new implementation seems ok, but a complete new type...?

Comment (panquez) [reply src/geode/mesh/core/tetrahedral_solid.cpp]: You are right, maybe new impl is better.

**Pull request merged**